### PR TITLE
Master sms provider feedback thc

### PR DESCRIPTION
**Pull request by Explicitapps**
Description of the issue/feature this PR addresses:

-Users doesn't know the real state of the sms sent. Only if there was no errors on iap-service, but that doesn't mean they where all sent to the contacts without problem.

Current behavior before PR:

-The only feedback is from iap-services as a return of the client request.

Desired behavior after PR is merged:

-Now there is a feedback coming from the provider to iap-services that we send back to client with a specific route.

Task: 2630217

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sms/models/sms_sms.py [hunk 1]
```diff
@@ -45,6 +45,23 @@ class SmsSms(models.Model):
         ('sms_optout', 'Opted Out'),
     ], copy=False)
 
+    request_id = fields.Char('request id', help="request id return by sms service")
```
Comment (fda-odoo) [inline addons/sms/models/sms_sms.py:48]: returned

File: addons/sms/models/sms_api.py [hunk 2]
```diff
@@ -52,8 +55,13 @@ def _send_sms_batch(self, messages):
...
         :raises: normally none
         """
+        url = self.get_base_url()
+
+        for sms in messages:
+            sms['webhook_url'] = url +'/sms/status'
+
         params = {
-            'messages': messages
+            'messages': messages,
```
Comment (fda-odoo) [inline addons/sms/models/sms_api.py:64]: 'messages': messages

**Review by fda-odoo — CHANGES_REQUESTED** [review]

**Review by fda-odoo — CHANGES_REQUESTED** [review]

**Review by tde-banana-odoo — CHANGES_REQUESTED** [review]
Validation as SM code maintainer: great to have the feedback for SMS, thanks for your PR ! However some cleaning is required, as state by various comments. Feel free to ask if anything is unclear.

Thanks !

**Comment by Explicitapps** [comment]
@tde-banana-odoo 

hello, I just pushed some changes. ;) 

SO... as about your concern about sms that are not deleted anymore :
They are. but later when IAP sends back the infos to the client. At that moment the state of the notifications and traces are changed accordingly and if the sms is delivered it' ll be deleted. If not it 'll stay with an error state and so the client could juste pick it and click on resend in the sms list view.
BUT... We need to change the state of all sms records on the _postprocess_iap_sent_sms method. SO the state does not stay in "outgoing" if we do not do that the cron will resend all those sms if the provider does not send back the state of all the messages in time.
Anyway... batches limited to 500 records at a time. and if some clients change that value in their code, there is a hard cap on IAP side of 5000. So if a batch of more than 5000 sms is send to IAP it will be rejected. ( to avoid timout problems )

I also saw that you changed the "delete_all" option by "unlink_sent" and  "unlink_failed" but where those option can be set ? I think there is no parameter in the settings or somewhere to define what you prefer

**Comment by DominikZians** [comment]
Hello @tde-banana-odoo,
we have resumed working on this PR. I made several changes and we would like your feedback/review before working on tests.

## Changes made to the IAP-API:
Before: a message was represented by a dict of the format:
```
{
    'res_id': string/int,
    'number': string,
    'content': string,
}
```
Now: a message is represented by a dict of the format:
```
{
    'request_uuid': string,
    'number': string,
    'content': string,
    'webhook_url': string (OPTIONAL),
}
```
The field `request_uuid` replaces `res_id` and serves as a UUID to identify messages client-side (where it is generated) but also IAP-side. The field `webhook_url` has been added, IAP will call this API when the delivery report (DLR) of the corresponding message is available. Once this call has happened the state of the message is not subject to change again.

## Changes to sms
- Changes needed to support the new IAP-API
- By default `sms.sms` are NOT unlinked anymore as soon as IAP answers with `'success'` for the given message. Only after receiving a successful DLR (`== 'delivered'`) the corresponding `sms.sms` IS ALWAYS unlinked (to discuss if it should always be unlinked).

## Changes to mass_mailing_sms
- Changes needed to support the new IAP-API
- Added the `sms_request_uuid` field to `mailing.trace` (which is the `request_uuid` of the corresponding `sms.sms`). This makes it possible to receive DLRs even if the corresponding `sms.sms` has been deleted. (This makes it possible to unlink an `sms.sms` as soon as IAP answers with `'success'` for the given message again, at least for sms send by the `mass_mailing_sms` module)

File: addons/sms/controller/main.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from odoo.http import request
+
+
+_logger = logging.getLogger(__name__)
+
+
+class SmsController(http.Controller):
+
+    @http.route('/sms/status', type='json', auth='public', csrf=False)
+    def update_sms_status(self, message):
```
[the remark this answers is not in the document]
Comment (DominikZians) [reply addons/sms/controller/main.py]: Wouldn't it be better for this route to accept a dict `messages` of the form:
```
{
    state0: [uuid00, uuid01, ...],
    state1: [uuid10, uuid11, ...],
    ...
}
```
Which would enable IAP to batch DLRs?
Imagine the client wants to send 1000 SMS, in the case where IAP is not able to transmit the batch to our providers we will make 1000 calls to this webhook to say that we were not able to deliver the sms messages!

**Comment by tde-banana-odoo** [comment]
Hello @DominikZians ,

Thanks for updating ! Will have a look when having some time, don't hesitate to ping me directly on discord if necessary. I also have a task in SM project to make the followup, not sure what is the status on your side (can be tested, ...) as the linked task in the PR does not contain much information. You can also ping me on the SM tracking task, it is assigned to you (see task 2560666).

Cheers !

**Comment by seb-odoo** [comment]
task was marked as done

**Pull request closed without merging**